### PR TITLE
Make function properties context-sensitive based on their return statements

### LINKED ISSUE #50687: Type inference fails with nested arrow function callback with block
# Bug Report

### 🔎 Search Terms

callback, block, inference, double

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about arrow functions

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAJwKYAdlwDwBUB8AFAPoQLAwDmAXIgN4BQii6AhsqwLYDOtuA3E0QRWAG1EAjVhADWtQmw48+ASkQBefIgXsuvRLjWbEYEJwmpkggL4raANzgwAJvRSooIZEmsMGaTDhCRmZFPVoARgAaIRFxKVl5Iy06d09vRFYNLSyAakQIxGsoxAB6UsRwGTA4AHcwBltBAKxgoTDlApjmOMlpOURiCWy3NHSkLOM8gqKS8sQEUQBPRGcYYGBLVEhURBhuRCgAC12qGrRXDsaVZoxWkJZdTujYsT7E7WTMkenosorFis1hstjs9gdjrtRP0FsBEAADVJjLxIAB06KK8Ou-CAA

### 💻 Code

```ts
function repro<T>(_config: {
  params: T;
  callback: (params: T) => (params: T) => number;
}): void { return }

repro({
  params: 1,
  callback: () => { return a => a + 1 }, // a is unknown
});
repro({
  params: 1,
  callback: _b => { return a => a + 1 }, // a is a number, only difference is the ignored outer param
});
repro({
  params: 1,
  callback: () => a => a + 1, // a is a number, only difference is the lack of `{ return ... }`
});
```

### 🙁 Actual behavior

The type of `a` ends up being `unknown`, but only if no parameter is declared and a block is used in the outer callback.

### 🙂 Expected behavior

The type of the inner callback parameter `a` should be known in all 3 cases.

### PR DESCRIPTION
**Pull request by Andarist**
fixes #50687

cc @Rugvip

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -17588,8 +17588,13 @@ namespace ts {
         }
 
         function hasContextSensitiveReturnExpression(node: FunctionLikeDeclaration) {
-            // TODO(anhans): A block should be context-sensitive if it has a context-sensitive return value.
```
Comment (Andarist) [inline src/compiler/checker.ts:17591]: Basically the fix implements this TODO comment 😉

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot perf test this faster
@typescript-bot test top100

**Comment by RyanCavanaugh** [comment]
Related -- #47599

**Comment by RyanCavanaugh** [comment]
@typescript-bot pack this

**Comment by Andarist** [comment]
@RyanCavanaugh all cases from the first post in [47599](https://github.com/microsoft/TypeScript/issues/47599#issue-1114250433) work OK since 4.7: [TS playground](https://www.typescriptlang.org/play?ts=4.7.4#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzCgggEkMAeAFQD4AKHAIwCsAueAbwCh4f4AHGDmDIwINrVRtUyALYMQMAJTwAvNXiUANN15g8AZ1lj4tAB5tKytfABuOLME4BfRWzsOA3J06FiZWly8-ILCouJW6gAM2kF6qIYyxvjWqAB0GDgAYlimIMC0is6KXj5EpBgBOjwCQiLGAPpQqlExugZGUs3waRnZufmFLiW+5ZVBNaEgBRxVQfBwGMgw+JFeQU6tPHEJSV09WTl5BUUeQA). Do you have any additional test cases for this PR based on that issue? Could that issue be already closed as solved based on this prepared playground?

**Comment by RyanCavanaugh** [comment]
@weswigham @ahejlsberg I remember from prior conversations that this approach is maybe not *quite* right (due to CFA in function bodies?), but I wasn't able to craft a demonstration of a problem. Thoughts?

**Comment by weswigham** [comment]
Iirc it's because the `isContextSensitive` logic isn't smart enough to be able to climb through all the branches/conditions in a block body to check if a return statement is, in a roundabout way, actually context sensitive. So it's more that this is incomplete rather than incorrect - but it was already incomplete by virtue of never marking return statements as context sensitive, so... 🤷

**Comment by nix6839** [comment]
> Hey @RyanCavanaugh, I've packed this into [an installable tgz](https://typescript.visualstudio.com/cf7ac146-d525-443c-b23c-0d58337efebc/_apis/build/builds/135059/artifacts?artifactName=tgz&fileId=9B9E7C827F6AF362B57D1037A86D3FCCF508F8EDBB34F60B1E03344AB433109902&fileName=/typescript-4.9.0-insiders.20220922.tgz). You can install it for testing by referencing it in your `package.json` like so:
> 
> ```
> {
>     "devDependencies": {
>         "typescript": "https://typescript.visualstudio.com/cf7ac146-d525-443c-b23c-0d58337efebc/_apis/build/builds/135059/artifacts?artifactName=tgz&fileId=9B9E7C827F6AF362B57D1037A86D3FCCF508F8EDBB34F60B1E03344AB433109902&fileName=/typescript-4.9.0-insiders.20220922.tgz"
>     }
> }
> ```
> 
> and then running `npm install`.
The above link has expired. How can I use this PR?

**Comment by jakebailey** [comment]
Those builds expire, but it's supposed to go and build an npm package / playground too. I'll rerun it.

@typescript-bot pack this

**Comment by nix6839** [comment]
```
function fn<ToInferred>(
  callback: (unused: string) => {
    key: ToInferred;
    keyCallback: (key: ToInferred) => number;
  },
) {}

fn(() => ({
  key: 5,
  keyCallback: (key) => {
    return key; // Success, key is number
  },
}));

fn((unused) => ({
  key: 5,
  keyCallback: (key) => {
    return key; // Error, key is unknown.
  },
}));
```
The above code is not working. Should I create new issue?

**Comment by Andarist** [comment]
I would say that yes, this might deserve a new issue - it doesn't exactly look like the one that is being fixed here.

**Comment by benjdlambert** [comment]
Hey - is there any update on this? Would be good to get this merged :pray:

**Comment by Andarist** [comment]
@RyanCavanaugh @weswigham any further thoughts on this one?

**Review by weswigham — APPROVED** [review]
Honestly, I've written this exact change 2 or 3 times since that TODO was first introduced, and given how limited return type inference is, I think this is good enough™️. Plus, any issue you could bring forward as block-specific and a good reason to forbid blocks, I can probably  rewrite into a blockless arrow using the comma operator and an iife, which we *should* already track context sensitivity through (spoiler: we don't, same as we don't across block statements), so any limitation in considering the block is probably already equivalently demonstrable today.

Unless @ahejlsberg feels really strongly and still has good reason to block this, I'm inclined to take it. (pending test results)

**Comment by weswigham** [comment]
@typescript-bot run dt
@typescript-bot test top100

**Pull request merged**